internal/ansible/ansible_manager.go

```diff
@@ -0,0 +1,248 @@
+	return &AnsibleManager{
+		dispatcherClient:                 dispatcherClient,
+		ansibleDispatcher:                dispatcher.NewAnsibleDispatcher(deviceID),
+		ansiblePlaybookConnectionOptions: playbookConnectionOptions,
+		ansiblePlaybookOptions:           playbookOptions,
+		playbook:                         playbookCmdParam,
+	}
+}
+
+func MissingAttributeError(attribute string, metadata map[string]string) error {
+	return fmt.Errorf(missingAttributeMsg(attribute, metadata))
+}
+
+func missingAttributeMsg(attribute string, metadata map[string]string) string {
+	return fmt.Sprintf("missing attribute %s in message metadata %v", attribute, metadata)
+}
+
+func invalidAttributeMsg(attribute string, value string) string {
+	return fmt.Sprintf("invalid attribute %s = %s in message", attribute, value)
+}
+
+func (a *AnsibleManager) HandlePlaybook(d *pb.Data) error {
+	// required fields
+	reqFields := &RequiredFields{}
+	payloadStr := string(d.Content)
+
+	found := false
+	dataDir := "/tmp"
+
+	var err error
+	responseTo := d.MessageId
+	if payloadStr == "" {
+		return fmt.Errorf("missing playbook string in message")
+	}
+	if reqFields.crcDispatcherCorrelationID, found = d.GetMetadata()[crcDispatcherAttribute]; !found {
+		return MissingAttributeError(crcDispatcherAttribute, d.GetMetadata())
+	}
+	if responseInterval, found := d.GetMetadata()[responseIntervalAttribute]; !found {
+		log.Warnf(missingAttributeMsg(responseIntervalAttribute, d.GetMetadata())+" Defaulting to "+strconv.Itoa(defaultResponseInterval), d.GetMetadata())
+		reqFields.responseInterval = defaultResponseInterval
+	} else {
+		reqFields.responseInterval, err = strconv.ParseUint(responseInterval, 10, 64)
+		if err != nil {
+			log.Warnf(invalidAttributeMsg(responseIntervalAttribute, d.GetMetadata()[responseIntervalAttribute]+" Defaulting to "+strconv.Itoa(defaultResponseInterval)))
+			reqFields.responseInterval = defaultResponseInterval
+		}
+	}
+	if reqFields.returnURL, found = d.GetMetadata()[returnURLAttribute]; !found {
+		return fmt.Errorf(missingAttributeMsg(returnURLAttribute, d.GetMetadata()))
+	}
+
+	playbookYamlFile = path.Join(dataDir, "ansible_playbook_"+d.MessageId+".yml")
+	defer os.Remove(playbookYamlFile)
+	err = os.WriteFile(playbookYamlFile, d.GetContent(), 0600)
+
+	if err != nil {
+		return fmt.Errorf("cannot create ansible playbook yaml file %s. Error: %v", playbookYamlFile, err)
+	}
+
+	/*
+	*******************************************
+	*  TODO : Verify playbook
+	*******************************************
+	 */
+
+	// required event for cloud connector
+	onStart := a.ansibleDispatcher.ExecutorOnStart(reqFields.crcDispatcherCorrelationID, "")
+	a.ansibleDispatcher.AddRunnerJobEvent(onStart)
+
+	executionCompleted := make(chan error)
+	playbookResults := make(chan *ansibleResults.AnsiblePlaybookJSONResults)
+
+	// execute
+	go execPlaybook(executionCompleted, playbookResults, playbookCmd, time.Duration(reqFields.responseInterval)*time.Second)
```

Comment:
Should we have a context to stop this in some way? Worried that get stuck the playbook and there is no way to cancel, etc..

Reply:
I mean, I'm worried that something waits forever on ansible, and playbook got stuck on device